File: app/src/main/java/cn/ifafu/ifafu/http/parser/LoginParser.java
package cn.ifafu.ifafu.http.parser;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import cn.ifafu.ifafu.data.Response;
import io.reactivex.Observable;
import io.reactivex.functions.Function;
import okhttp3.ResponseBody;

public class LoginParser implements Function<ResponseBody, Response<String>> {

    /**
     * @param html 网页信息
     * @return {@link Response#SUCCESS} 登录成功 body = user only with name
     *         {@link Response#FAILURE} 信息错误 msg = return msg
     *         {@link Response#ERROR}   服务器错误  msg = error msg
     */
    public Response<String> parse(String html) {
        Document doc = Jsoup.parse(html);
        Element ele = doc.getElementById("xhxm");
        if (ele != null) {
            String name = ele.text().replace("同学", "");
            return Response.success(name);
        }
        Elements script = doc.select("script[language=javascript]");
        if (script.size() < 2) {
            return Response.error("网络异常");
        } else {
            String s = getAlertString(script.get(1).html());
            if (s.contains("用户名") || s.contains("密码") || s.contains("验证码")) {
                return Response.failure(s);
            } else if (doc.text().contains("ERROR")) {
                return Response.error("教务系统又双叒崩溃了");
            } else {
                return Response.error("网络异常");
            }
        }
    }

    private String getAlertString(String text) {
        Pattern p = Pattern.compile("alert\\('.*'\\);");
        Matcher m = p.matcher(text);
        if (m.find()) {
            String s = m.group();
            return s.substring(7, s.length() - 3);
        }
        return "";
    }

    @Override
    public Response<String> apply(ResponseBody responseBody) throws Exception {
        return parse(responseBody.string());
    }
}

File: app/src/main/java/cn/ifafu/ifafu/http/parser/LoginParser.kt
package cn.ifafu.ifafu.http.parser

import cn.ifafu.ifafu.data.Response
import cn.ifafu.ifafu.http.exception.VerifyErrorException
import io.reactivex.Observable
import io.reactivex.ObservableSource
import io.reactivex.ObservableTransformer
import okhttp3.ResponseBody
import org.jsoup.Jsoup
import java.util.regex.Pattern

class LoginParser : ObservableTransformer<ResponseBody, Response<String>> {

    /**
     * 登录页面解析
     *
     * @param html 网页信息
     * @return [Response.SUCCESS] 登录成功 body = user only with name
     * @return [Response.FAILURE] 信息错误 msg = return msg
     * @return [Response.ERROR]   服务器错误  msg = error msg
     */
    private fun parse(html: String): Response<String> {
        val doc = Jsoup.parse(html)
        val ele = doc.getElementById("xhxm")
        if (ele != null) {
            val name = ele.text().replace("同学", "")
            return Response.success(name)
        }
        val script = doc.select("script[language=javascript]")
        return if (script.size < 2) {
            Response.error("未知错误")
        } else {
            val s = getAlertString(script[1].html())
            if (s.contains("用户名") || s.contains("密码")) {
                Response.failure(s)
            } else if (doc.text().contains("ERROR")) {
                Response.error("教务系统又双叒崩溃了")
            } else if (s.contains("验证码")) {
                throw VerifyErrorException()
            } else {
                Response.error("未知错误")
            }
        }
    }

    private fun getAlertString(text: String): String {
        val p = Pattern.compile("alert\\('.*'\\);")
        val m = p.matcher(text)
        if (m.find()) {
            val s = m.group()
            return s.substring(7, s.length - 3)
        }
        return ""
    }

    override fun apply(upstream: Observable<ResponseBody>): ObservableSource<Response<String>> {
        return upstream.map { responseBody -> parse(responseBody.string()) }
    }
}
